feat: make persistent-others similar to the new persistent-apps

ed9w2in6 opened the pull request:
This is backward compatible via conversion function provided for `coercedTo`.

Fixes: #968, #982, #1398

The old logic is as follows:  if ba

For this configuration:
```
  persistent-others = [
    ./.
    "/Volumes"
    { file = "/Volumes"; }
    "/Users/${username}/Desktop/this_is_a_file"
    { file = "/Users/${username}/Desktop/this_is_a_file"; }
    "/Users/${username}/.local/share/chezmoi/__.emacs.d"
    { folder = { path = "/Users/${username}/.local/share/chezmoi/__.emacs.d"; showas = "grid"; }; }
    "/Users/${username}/.emacs.d/"
    "/Users/${username}/.emacs.d"
    { folder = "/Users/${username}/Downloads"; }
  ];
```

In my computer, it wiill result in this:
![image](https://github.com/user-attachments/assets/155d1c78-10bd-447b-b5a1-e6e385abc4af)
- `./.` :: folder since `./.` is converted to a path like `/path/to/folder` so by the backward compatible logic it is a folder.
- `/Volumes` :: a folder
-

ed9w2in6 commented:
Conflicts with:
https://github.com/nix-darwin/nix-darwin/pull/1223
https://github.com/nix-darwin/nix-darwin/pull/1078
https://github.com/nix-darwin/nix-darwin/pull/1004

ed9w2in6 commented:
Rendered updated example in the documentation:
<img width="1211" alt="image" src="https://github.com/user-attachments/assets/2c4bdf4e-7b8d-456c-9aa9-40f2eefef122" />

ed9w2in6 commented:
May I ask if the tests are needed to be updated manually?
https://github.com/nix-darwin/nix-darwin/blob/master/tests/system-defaults-write.nix

Enzime commented:
Yeah you'll need to update the files in this folder https://github.com/nix-darwin/nix-darwin/tree/master/tests/fixtures/system-defaults-write

ed9w2in6 commented:
@Enzime I am not sure how to add a test case for paths while not being brittle to any minor changes to the test environment.  So I did NOT add a test case for paths right now, added a comment on this started with `# TODO:...`.

I have tested it on my own computer though.

ed9w2in6 commented:
@Enzime Sorry for pinging, may I ask if the tests are good enough?

ed9w2in6 commented:
rebased to master after system activation merged

ed9w2in6 commented:
@Enzime 
May I ask if this is good to be merged?

I understand that previously there were on going plans on refactoring to system activation.
Now that is already done, I had already rebased, adjusted, and tested my commits since then.
They should be good for review.

This is not a big change and should affect Dock only so there shouldn't be too much to worry.
Though if there are any concerns please let me know.
Personally I wish to remove this from my task list as soon as possible so I am keen to help.

Thank you for your time on maintaining this project again.

phinze commented:
Hi there 👋🏻 - Just chiming in to say I've just [tested this on my config](https://github.com/phinze/nix-config/commit/8fd42a8ca96886ab029a5b8679925815178f6b00) and it's working well for my use case. Thanks for your work on this @ed9w2in6!

Samasaur1 commented:
Thank you! (and apologies for the delay)
